Security_Engine/Create/CameraFieldOfView.cs

```diff
@@ -0,0 +1,364 @@
+            {
+                conePoints.AddRange(coneLine.ControlPoints());
+            }
+
+            //cone boundary points
+            List<Point> intersectPoints = new List<Point>();
+            intersectPoints.Add(conePoints[1]);
+            intersectPoints.Add(conePoints[conePoints.Count - 2]);
+
+            //points that intersect with obstacles
+            List<Polyline> intersectObstacles = new List<Polyline>();
+            foreach (Polyline obstacle in obstacles)
+            {
+                Polyline projObstacle = obstacle.Project(cameraPlane);
+                List<Polyline> intersection = projObstacle.BooleanIntersection(cameraConePolyline);
+                if (intersection.Count != 0)
+                {
+                    intersectObstacles.Add(projObstacle);
+                    foreach (Polyline intersectPolyline in intersection)
+                    {
+                        List<Point> points = intersectPolyline.ControlPoints;
+                        intersectPoints.AddRange(points);
+                    }
+                }
+            }
+            intersectPoints = intersectPoints.CullDuplicates();
+
+            //create and sort extended ray lines passing through intersected points
+            List<Line> rayLines = new List<Line>();
+            foreach (Point point in intersectPoints)
+            {
+                Line line = BH.Engine.Geometry.Create.Line(cameraLocation, point);
+                Line rayLine = line.Extend(0, radius - line.Length());
+                rayLines.Add(rayLine);
+            }
+            rayLines = rayLines.CullDuplicateLines();
+            rayLines = rayLines.OrderBy(x => x.Angle(rayLines[0])).ToList();
+
+            //split ray lines and find visible line
+            List<Dictionary<Line, Polyline>> linesDict = new List<Dictionary<Line, Polyline>>();
+            List<Line> visibleLines = new List<Line>();
+            foreach (Line rayLine in rayLines)
+            {
+                Line visibleLine = rayLine.VisibleLine(intersectObstacles);
+                visibleLines.Add(visibleLine);
+                linesDict.Add(visibleLine.LineObstacleDictionary(intersectObstacles));
+            }
+
+            //create points chain
+            List<Point> pointsChain = linesDict.PointsChain(cameraLocation, radius);
+
+            //create cone
+            Arc coneArc = cameraCone.Curves[1] as Arc;
+            PolyCurve cameraViewPolyCurve = pointsChain.CameraViewPolyCurve(cameraLocation, radius, coneArc);
+
+            return cameraViewPolyCurve;
+        }
+
+        /***************************************************/
+
+        [Description("Returns the polycurve camera view cone.")]
+        [Input("cameraDevice", "The CameraDevice object to compute the camera view cone for.")]
+        [Output("conePolyCurve", "PolyCurve object that represents camera view cone.")]
+        public static PolyCurve CameraViewCone(this CameraDevice cameraDevice)
```

Comment:
This could be a `Query` method named `ViewCone` I guess?

Reply:
Good idea, will do it 👍